manager/orchestrator/replicated/replicated_test.go

```diff
@@ -424,3 +425,133 @@ func TestReplicatedScaleDown(t *testing.T) {
 		}
 	})
 }
+
+func TestInitialization(t *testing.T) {
+	ctx := context.Background()
+	s := store.NewMemoryStore(nil)
+	assert.NotNil(t, s)
+	defer s.Close()
+
+	orchestrator := NewReplicatedOrchestrator(s)
+	defer orchestrator.Stop()
+
+	watch, cancel := state.Watch(s.WatchQueue())
+	defer cancel()
+
+	s1 := &api.Service{
+		ID: "id1",
+		Spec: api.ServiceSpec{
+			Annotations: api.Annotations{
+				Name: "name1",
+			},
+			Mode: &api.ServiceSpec_Replicated{
+				Replicated: &api.ReplicatedService{
+					Replicas: 2,
+				},
+			},
+		},
+	}
+
+	err := s.Update(func(tx store.Tx) error {
+		assert.NoError(t, store.CreateService(tx, s1))
+
+		nodes := []*api.Node{
+			{
+				ID: "node1",
+				Spec: api.NodeSpec{
+					Annotations: api.Annotations{
+						Name: "name1",
+					},
+					Availability: api.NodeAvailabilityActive,
+				},
+				Status: api.NodeStatus{
+					State: api.NodeStatus_READY,
+				},
+			},
+			{
+				ID: "node2",
+				Spec: api.NodeSpec{
+					Annotations: api.Annotations{
+						Name: "name2",
+					},
+					Availability: api.NodeAvailabilityActive,
+				},
+				Status: api.NodeStatus{
+					State: api.NodeStatus_READY,
+				},
+			},
+		}
+		for _, node := range nodes {
+			assert.NoError(t, store.CreateNode(tx, node))
+		}
+
+		// 2 rejected tasks are in store before orchestrator starts
+		tasks := []*api.Task{
+			{
+				ID:           "task1",
+				Slot:         1,
+				DesiredState: api.TaskStateReady,
+				Status: api.TaskStatus{
+					State: api.TaskStateRejected,
+				},
+				ServiceAnnotations: api.Annotations{
+					Name: "task1",
+				},
+				ServiceID: "id1",
+				NodeID:    "node1",
+			},
+			{
+				ID:           "task2",
+				Slot:         2,
+				DesiredState: api.TaskStateReady,
+				Status: api.TaskStatus{
+					State: api.TaskStateNew,
+				},
+				ServiceAnnotations: api.Annotations{
+					Name: "task2",
+				},
+				ServiceID: "id1",
+				NodeID:    "",
+			},
+		}
+		for _, task := range tasks {
+			assert.NoError(t, store.CreateTask(tx, task))
+		}
+
+		return nil
+	})
+	assert.NoError(t, err)
+
+	// Start the orchestrator.
+	go func() {
+		assert.NoError(t, orchestrator.Run(ctx))
+	}()
+
+	// wait for orchestration to finish
+	done := false
+	for !done {
+		select {
+		case <-watch:
+		case <-time.After(time.Second):
+			done = true
```

Comment:
There are the events for slot 1. `initTask()` tries to restart the task, moving it DesiredState from `READY` to `RUNNING`, then `tick()` would move DesiredState to `SHUTDOWN`, and start a new task to replace it. So it's doing the right thing. There is a lot of implementation details in the event sequence, and they may change when implementation changes. That's why I try to check the end result instead of the events. 

```
event: {&Task{ID:task1,Meta:Meta{Version:Version{Index:0,},CreatedAt:(timestamp: nil google_protobuf.Timestamp),UpdatedAt:(timestamp: nil google_protobuf.Timestamp),},Spec:TaskSpec{Runtime:<nil>,Resources:nil,Restart:nil,Placement:nil,LogDriver:nil,Networks:[],ForceUpdate:0,},ServiceID:id1,Slot:1,NodeID:node1,Annotations:Annotations{Name:,Labels:map[string]string{},},ServiceAnnotations:Annotations{Name:task1,Labels:map[string]string{},},Status:TaskStatus{Timestamp:(timestamp: nil google_protobuf.Timestamp),State:REJECTED,Message:,Err:,RuntimeStatus:<nil>,PortStatus:nil,},DesiredState:RUNNING,Networks:[],Endpoint:nil,LogDriver:nil,} []}

event: {&Task{ID:task1,Meta:Meta{Version:Version{Index:0,},CreatedAt:(timestamp: nil google_protobuf.Timestamp),UpdatedAt:(timestamp: nil google_protobuf.Timestamp),},Spec:TaskSpec{Runtime:<nil>,Resources:nil,Restart:nil,Placement:nil,LogDriver:nil,Networks:[],ForceUpdate:0,},ServiceID:id1,Slot:1,NodeID:node1,Annotations:Annotations{Name:,Labels:map[string]string{},},ServiceAnnotations:Annotations{Name:task1,Labels:map[string]string{},},Status:TaskStatus{Timestamp:(timestamp: nil google_protobuf.Timestamp),State:REJECTED,Message:,Err:,RuntimeStatus:<nil>,PortStatus:nil,},DesiredState:SHUTDOWN,Networks:[],Endpoint:nil,LogDriver:nil,} []}

event: {&Task{ID:ainr2bk0mkxccn27i89lg7tcu,Meta:Meta{Version:Version{Index:0,},CreatedAt:(timestamp: nil google_protobuf.Timestamp),UpdatedAt:(timestamp: nil google_protobuf.Timestamp),},Spec:TaskSpec{Runtime:<nil>,Resources:nil,Restart:nil,Placement:nil,LogDriver:nil,Networks:[],ForceUpdate:0,},ServiceID:id1,Slot:1,NodeID:,Annotations:Annotations{Name:,Labels:map[string]string{},},ServiceAnnotations:Annotations{Name:name1,Labels:map[string]string{},},Status:TaskStatus{Timestamp:2017-02-01T00:07:11.897776845Z,State:NEW,Message:created,Err:,RuntimeStatus:<nil>,PortStatus:nil,},DesiredState:READY,Networks:[],Endpoint:&Endpoint{Spec:nil,Ports:[],VirtualIPs:[],},LogDriver:nil,} []}
```